Source/Core/VideoCommon/PostProcessing.cpp

```diff
@@ -2,87 +2,213 @@
+	const std::vector<std::string> search_extensions = { ".glsl" };
+	std::vector<std::string> result;
+	std::vector<std::string> paths;
 
-PostProcessingShaderImplementation::~PostProcessingShaderImplementation()
-{
-	m_timer.Stop();
+	// main folder
+	paths = DoFileSearch(search_extensions, search_dirs, false);
+	for (const std::string& path : paths)
+	{
+		std::string filename;
+		if (SplitPath(path, nullptr, &filename, nullptr))
+		{
+			if (std::find(result.begin(), result.end(), filename) == result.end())
+				result.push_back(filename);
+		}
+	}
+
+	// folders/sub-shaders
+	paths = FindSubdirectories(search_dirs, false);
+	for (const std::string& dirname : paths)
+	{
+		// find sub-shaders in this folder
+		size_t pos = dirname.find_last_of(DIR_SEP_CHR);
+		if (pos != std::string::npos && (pos != dirname.length() - 1))
+		{
+			std::string shader_dirname = dirname.substr(pos + 1);
+			std::vector<std::string> sub_paths = DoFileSearch(search_extensions, { dirname }, false);
+			for (const std::string& sub_path : sub_paths)
+			{
+				std::string filename;
+				if (SplitPath(sub_path, nullptr, &filename, nullptr))
+				{
+					// Remove /main for main shader
+					std::string name = (!strcasecmp(filename.c_str(), "main")) ? (shader_dirname) : (shader_dirname + DIR_SEP + filename);
+					if (std::find(result.begin(), result.end(), filename) == result.end())
+						result.push_back(name);
+				}
+			}
+		}
+	}
+
+	// sort lexicographically
+	std::sort(result.begin(), result.end());
+	return result;
 }
 
-std::string PostProcessingShaderConfiguration::LoadShader(std::string shader)
+bool PostProcessingShaderConfiguration::LoadShader(const std::string& sub_dir, const std::string& name)
 {
-	// Load the shader from the configuration if there isn't one sent to us.
-	if (shader == "")
-		shader = g_ActiveConfig.sPostProcessingShader;
-	m_current_shader = shader;
+	// clear all state
+	m_shader_name = name;
+	m_shader_source.clear();
+	m_options.clear();
+	m_render_passes.clear();
+	m_any_options_dirty = false;
+	m_compile_time_constants_dirty = false;
+	m_requires_depth_buffer = false;
 
-	const std::string sub_dir = (g_Config.iStereoMode == STEREO_ANAGLYPH) ? ANAGLYPH_DIR DIR_SEP : "";
+	// special case: default shader, no path, use inbuilt code
+	if (name.empty())
+	{
+		m_shader_source = s_default_shader;
+		return ParseConfiguration(sub_dir, "");
+	}
 
-	// loading shader code
-	std::string code;
-	std::string path = File::GetUserPath(D_SHADERS_IDX) + sub_dir + shader + ".glsl";
+	// This is kind of horrifying, but we only want to load one shader, in a specific order.
+	// (if there's an error, this class is left in an unknown state)
+	std::string dirname;
+	std::string filename;
+	bool found = false;
+	bool result = false;
 
-	if (shader == "")
+	// Try old-style shaders first, but completely skip if the name is a sub-shader
+	if (name.find_first_of(DIR_SEP_CHR) == std::string::npos)
```

Comment:
this can just be `name.find`, find_first_of is only useful when looking for the first of a sequence of characters.